src/contract_client.rs

```diff
@@ -85,7 +95,312 @@ impl From<RejectReason> for ViewError {
+                AccountTransactionEffects::ContractInitialized { data } => {
+                    let contract_client =
+                        ContractClient::new(self.client, data.address, data.init_name);
+                    Ok((contract_client, data.events))
+                }
+                AccountTransactionEffects::None {
+                    transaction_type: _,
+                    reject_reason,
+                } => Err(ContractInitError::Failed(reject_reason)),
+                _ => mk_error(
+                    "Expected smart contract initialization status, but did not receive it.",
+                ),
+            },
+            crate::types::BlockItemSummaryDetails::AccountCreation(_) => mk_error(
+                "Expected smart contract initialization status, but received account creation.",
+            ),
+            crate::types::BlockItemSummaryDetails::Update(_) => mk_error(
+                "Expected smart contract initialization status, but received chain update \
+                 instruction.",
+            ),
+        }
+    }
+
+    /// Wait until the transaction is finalized or until the timeout has elapsed
+    /// and return the result.
+    pub async fn wait_for_finalization_timeout(
+        self,
+        timeout: std::time::Duration,
+    ) -> Result<(ContractClient<Type>, Vec<ContractEvent>), ContractInitError> {
+        let result = tokio::time::timeout(timeout, self.wait_for_finalization()).await;
+        match result {
+            Ok(r) => r,
+            Err(_elapsed) => Err(ContractInitError::Query(QueryError::RPCError(
+                RPCError::CallError(tonic::Status::deadline_exceeded(
+                    "Deadline waiting for result of transaction is exceeded.",
+                )),
+            ))),
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+/// An error that may occur when attempting to dry run a new instance creation.
+pub enum DryRunNewInstanceError {
+    #[error("Dry run succeeded, but contract initialization failed due to {0:#?}.")]
+    Failed(RejectReason),
+    #[error("Dry run failed: {0}")]
+    DryRun(#[from] dry_run::DryRunError),
+}
+
 impl<Type> ContractClient<Type> {
+    /// Initialize a new smart contract instance and create a client as a
+    /// result. In contrast to
+    /// [`dry_run_new_instance_raw`](Self::dry_run_new_instance_raw) this
+    /// automatically serializes the provided parameter.
+    pub async fn dry_run_new_instance<P: contracts_common::Serial, E>(
+        client: Client,
+        sender: AccountAddress,
+        mod_ref: ModuleReference,
+        name: &str,
+        amount: Amount,
+        parameter: &P,
+    ) -> Result<ContractInitBuilder<Type>, E>
+    where
+        E: From<NewContractNameError>
+            + From<RejectReason>
+            + From<dry_run::DryRunError>
+            + From<v2::QueryError>
+            + From<ExceedsParameterSize>, {
```

Comment:
I don't think this works with something like `anyhow`.

Reply:
It does not, no, because RejectReason.